### ISSUE TITLE
Clingoext ground_program 

### ISSUE THREAD

**Issue by jorgefandinno** [jorgefandinno/eclingo#8]
clingoext ground_program has been replaced by their clingox counterpart. Can we remove it?

https://github.com/jorgefandinno/eclingo/blob/4c610d272584d0981e4af7c33299563ffc37bdb7/eclingo/util/clingoext.py#L86

**Comment by lute47lillo** [comment]
I have been working on this issue and I have found a couple things that need to be addressed:

1. When a program is grounded it gives a ground program such as: ['__x1.', '__x2.', 'u_a.', 'u_b.'] instead of just returning ['u_a.', 'u_b.']. I do not understand where that '__x1' comes from (I guess they are the variable placeholders), I think it has to do with the grounding process. I've checked the clingox documentation but I'm confused here. Maybe we just need to rewrite the expected outputs (?) Linking into the next problem....

2. On file test_pretty.py - test double negation works but it was failing because of how the asserting statements were being parsed, this has to do with how the ground_program of clingox works instead of using GroundProgram(). Here again, the ground program returns something like ["{d}.", "e :- not __x2.", "__x2 :- not d."] Instead of returning the normal x_2 used to ground the program. I don't understand that conversion. But in this case, there isn't anything added.

3. Program is not 'iterable' and that is causing almost 40 tests to fail.

4. there are some test_clingoext.py tests that aren't passing, but from our conversation yesterday I'm not worried about them yet.

If I can get some clarification on the __x1 appearance will help in order to fix the whole issue. 

I have 49 tests passed and 57 failed as of right now. I'm pushing the changes I made into a new branch lute/GroundProgramFix

**Comment by jorgefandinno** [comment]
Gringo introduces auxiliary atoms during grounding. In particular to deal with double negation.
For instance, if you have a program containing

p :- not not q.

It will be translated as something like

p :- not __x1.
__x1 :- not q.

In reality, the new introduced atom do not have a name (in clingo internals everything is a number). 
So we actually need to invent a name to be able to show them. 
If I remember correctly eclingo uses x1, x2,.. while the new library uses __x1, __x2 to avoid conflict if somebody uses x1, x2 themselves.

Does this help?

**Comment by lute47lillo** [comment]
So, should I re-write the tests again and giving the expected string the correspondent __x1 atoms instead of x_1 (which is what eclingo was using)? Or should they just be ommited on the tests?

https://github.com/jorgefandinno/eclingo/blob/931aebba2d755449d9b62584572172a3aa5d2c92/tests/test_grounder.py#L97-L101

In this case, the ground program returns: ['__x1.', 'not_sn_u_a.', 'u_b :- k_not_sn_u_a.', '{k_not_sn_u_a}.']. 
Then for the expected string we are actually missing the auxiliar atom __x1 for the case -a literal inside the subjective literal &k{not -a}, right? 
Does it mean that the ground program is wrong because __x1 is actually not part of the world view, or is it right and therefore the expected answer is what should be changed?

**Comment by jorgefandinno** [comment]
Rewrite the tests, but only those atoms.
Let me know if something else changed.

**Comment by lute47lillo** [comment]
Okey, so for some tests I changed the x_3 for __x3, and they passed. However, for some other tests I had to add __x1 to the expected since the ground program was given them. As you said, this is the grounding, once it goes to the solver, they are ignored.
The tests that needed this changes were:
test_epistemic_atom_with_both_negations01b
test_epistemic_atom_with_both_negations02
test_parameters
test_show01
test_show02
test_show03
test_epistemic_with_variables04
test_epistemic_with_variables05

**Comment by jorgefandinno** [comment]
Try just to remove all the strings starting with __x in the result of the ground program.
These are just an implementation detail.

> Okey, so for some tests I changed the x_3 for __x3, and they passed. However, for some other tests I had to add __x1 to the expected since the ground program was given them. As you said, this is the grounding, once it goes to the solver, they are ignored. The tests that needed this changes were: test_epistemic_atom_with_both_negations01b test_epistemic_atom_with_both_negations02 test_parameters test_show01 test_show02 test_show03 test_epistemic_with_variables04 test_epistemic_with_variables05

Ok, add it. That is no problem, as long as you don't have to modify anything that is not __x

**Comment by lute47lillo** [comment]
Okey, I have pushed the most recent changes to branch lute/#8GroundProgram.
There are some things that I'd like you to take a look at:

1. test_clingoext.py have been ignored for now, so I haven't changed anything there.
2. On the clingoext.py file I have made a change where I have changed:
https://github.com/jorgefandinno/eclingo/blob/e4c3555e259668d444d918033bc776a41ab67f8f/src/eclingo/util/clingoext.py#L119-L129
Therefore solving the issue given as 'Program" object is not iterable, which by the way is the most common error I've been fixing so far when trying to get rid of GroundProgram().
The problem now is that most tests from test_eclingo.py fail due to: **RuntimeError: Program is unsatisfiable.**
I have thought 2 ideas of why this is happening:

- The fix that I made and posted on the lines above shouldn't have been made, and that is why is giving unsatisfiable issues
_OR_
- The tests are actually unsatisfiable, which I highly doubt

However, since the introduction of the modal operator M there are now some tests that failed like this:
AssertionError: Lists differ: [['&m{-a}']] != [['&k{-a}']]
Where the world view gives that it is believed that -a is true -> &m{-a}, and the test expects that it is known that -a is true -> &k{-a}

I've pushed the changes, so the tests are going to run and you can see this exact FAILED tests to check it.
I think it'll be good if you can check it, I have some questions before I try to fix it all. Thank you!

**Comment by jorgefandinno** [comment]
Thank you, I will take a look.

**Comment by jorgefandinno** [comment]
I think the problem is in the new implementation of ```clingoext.add_to```

I reimplemented it on branch ```jorge/clingoext_add_to```

All test are passing.

Try to reapply your changes starting from this branch.
- Keep both ```ground_program``` and ```new_ground_program``` for now
- Replace one reference of  ```ground_program``` by```new_ground_program``` at the time
- Leave changes that break a test to last (including those in ```test_clingoext.py```), but rewrite the ```x1,...``` by ```__x1,...```

Once you have this let send it to me for revision.

**Comment by lute47lillo** [comment]
I have kept both ground_program and new program for now. I have changed most references of ground_program for new_ground_program except for test_clingoext.py and test_show01b and tests_show02b.

I believe that those _test_show and the tests from test_clingoext.py_ are failing for the same reason:
https://github.com/jorgefandinno/eclingo/blob/e0c71a1a5543168e2de89ee8649aee340f67b77c/tests/test_show.py#L57-L61

The expected program (from test_show) is given in that notation, however clingox Rules class is:

**class Rule(NamedTuple):
    '''
    Ground representation of disjunctive and choice rules.
    '''
    choice: bool
    head: Sequence[Atom]
    body: Sequence[Literal]**

Therefore, those test might need to be change to the new Rule class instead of the old ClingoRule class.

All changes have been pushed into a new lute/ground_programFix branch.
The rest of tests pass after changing references.
